### PR TITLE
Modified Battle Gloves' critical effect to be deterministic

### PR DESCRIPTION
**Pull request by Marsunpaisti**

### CODE REVIEW

File: wurst/_items/BattleglovesCritical.wurst [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    ..setItemAbility(false)
+    ..setHeroAbility(false)
+    ..presetDamageMultiplier(lvl -> MULTIPLIER)
+    ..presetDamageBonus(lvl -> 0)
+    ..presetChancetoCriticalStrike(lvl -> 100)
+    ..presetChancetoEvade(lvl -> 0)
+    ..presetCooldown(lvl -> 0)
+    ..presetNeverMiss(lvl -> true)
+
+@compiletime function createHiddenSpellbook()
```
Comment (jonathanmclaus) [inline wurst/_items/BattleglovesCritical.wurst]: Could you create a compiletime function that takes in a list of ability IDs and creates a spellbook containing those IDs. That way we could reuse this for other hidden abilities.

Comment (Marsunpaisti) [reply wurst/_items/BattleglovesCritical.wurst]: Created

File: wurst/_items/BattleglovesCritical.wurst [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    function onEnemyAttackDamage()
+        if wearer.hasAbility(SPELLBOOK_ABILITY_ID)
+            wearer.removeAbility(SPELLBOOK_ABILITY_ID)
+            this.lastCritTimestamp = this.baseClock
+        else
+            this.baseClock += COOLDOWN_SKIP_PER_ATTACK
+
+    ondestroy
+        print("Instance removed")
```
Comment (jonathanmclaus) [inline wurst/_items/BattleglovesCritical.wurst]: Remove print messages.

File: wurst/_items/BattleglovesCritical.wurst [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        if not instances.has(u)
+            instances.put(u, new BattleGlovesInstance(u))
+
+function onDropItem()
+    if GetManipulatedItem().getTypeId() == ITEM_BATTLE_GLOVES
+        let u = GetManipulatingUnit()
+        if instances.has(u)
+            destroy instances.getAndRemove(u)
+
+IterableMap<unit, BattleGlovesInstance> instances = new IterableMap<unit, BattleGlovesInstance>()
```
Comment (jonathanmclaus) [inline wurst/_items/BattleglovesCritical.wurst]: I'd move this up above `onPickupItem`, so it's defined before functions that reference it. Makes it more readable.

File: wurst/_items/BattleglovesCritical.wurst [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import DamageType
+import ID
+import AbilityObjectCreationUtils
+
+let CRITICAL_ABILITY_ID = compiletime(ABIL_ID_GEN.next())
+let SPELLBOOK_ABILITY_ID = compiletime(createHiddenSpellbookWithAbilities(toRawCode(CRITICAL_ABILITY_ID)))
+let COOLDOWN = 6.
+let COOLDOWN_SKIP_PER_ATTACK = 1.
+let MULTIPLIER = 1.6
+let CLOCK_TICK_INTERVAL = 0.05
```
Comment (jonathanmclaus) [inline wurst/_items/BattleglovesCritical.wurst]: Why not `ANIMATION_PERIOD`?

File: wurst/_items/BattleglovesCritical.wurst [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            let victim = GetTriggerUnit()
+            let attacker = GetEventDamageSource()
+            if instances.has(attacker) and not victim.isType(UNIT_TYPE_STRUCTURE) and not victim.getOwner().isAllyOf(attacker.getOwner())
+                instances.get(attacker).onEnemyAttackDamage()
+
+    doPeriodically(CLOCK_TICK_INTERVAL) cb ->
+        for i in instances
+            instances.get(i).updateTick()
+
+    EventListener.add(EVENT_PLAYER_UNIT_PICKUP_ITEM) ->
```
Comment (jonathanmclaus) [inline wurst/_items/BattleglovesCritical.wurst]: I would factor out the state in these callbacks then pass the unit and item to the respective functions.

File: wurst/_items/BattleglovesCritical.wurst [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    ..presetDamageMultiplier(lvl -> MULTIPLIER)
+    ..presetDamageBonus(lvl -> 0)
+    ..presetChancetoCriticalStrike(lvl -> 100)
+    ..presetChancetoEvade(lvl -> 0)
+    ..presetCooldown(lvl -> 0)
+    ..presetNeverMiss(lvl -> true)
+
+class BattleGlovesInstance
+    private unit wearer
+    private real lastCritTimestamp
```
Comment (jonathanmclaus) [inline wurst/_items/BattleglovesCritical.wurst]: I think you don't need this. Just check `baseClock > COOLDOWN` and reset it to `0`.

File: wurst/_items/BattleglovesCritical.wurst [hunk 7]
```diff
@@ -0,0 +1,89 @@
+package BattleglovesCritical
```
Comment (jonathanmclaus) [inline wurst/_items/BattleglovesCritical.wurst]: `Battle~g~GlovesCritical`

Change this and filename.

File: wurst/_items/BattleglovesCritical.wurst [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+class BattleGlovesInstance
+    private unit wearer
+    private real lastCritTimestamp
+    private real baseClock
+
+    construct(unit wearer)
+        this.wearer = wearer
+        this.baseClock = 0
+        this.lastCritTimestamp = 0
```
Comment (jonathanmclaus) [inline wurst/_items/BattleglovesCritical.wurst]: Remove this and the definition of the field in L32.

**Pull request merged**